### PR TITLE
Unroll linetext2's tail recursion to avoid stack level too deep

### LINKED ISSUE #608: Stack level too deep in smtpserver v1.0.7
I occasionally see a stack level too deep error in the smtpserver on line :570 I'm not exactly what causes this, because once the process is restarted, the upstream server re-sends the message that was in transit at failure time and everything succeeds at that point.

The smtpserver process is receiving about 15 emails/second when this crash happens

```
vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/smtpserver.rb:570:in `process_data_line': stack level too deep (SystemStackError)
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/smtpserver.rb:196:in `receive_line'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:63:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	 ... 6885 levels...
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/em/protocols/linetext2.rb:64:in `receive_data'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/eventmachine.rb:187:in `run_machine'
	from vendor/bundle/ruby/2.2.0/gems/eventmachine-1.0.7/lib/eventmachine.rb:187:in `run'
```

Any idea what might be causing this, or what I can do to work around it?

### PR DESCRIPTION
**Pull request by kjg**
Fixes #608 (hopefully)

### CODE REVIEW

**Comment by kjg** [comment]
@sodabrew All tests pass with this change, what do you think?

File: lib/em/protocols/linetext2.rb [hunk 1]
```diff
@@ -52,45 +52,51 @@ def receive_data data
...
-        if @lt2_mode == :lines
-          if ix = data.index( @lt2_delimiter )
-            @lt2_linebuffer << data[0...ix]
-            ln = @lt2_linebuffer.join
-            @lt2_linebuffer.clear
-            if @lt2_delimiter == "\n"
-              ln.chomp!
+        remaining_data = data
+
+        while remaining_data.length > 0
```
Comment (sodabrew) [inline lib/em/protocols/linetext2.rb]: use .bytesize (number of raw bytes) instead of .length (number of encoded characters)

Comment (sodabrew) [inline lib/em/protocols/linetext2.rb]: Oh, the prior code was all .length too. Hmm.

Comment (kjg) [inline lib/em/protocols/linetext2.rb]: If you'd like I can make that change here or in a separate PR, just let me know which. Might as well make other improvements while we're looking at it, right?

Comment (kjg) [inline lib/em/protocols/linetext2.rb]: I guess using length mirrors the use of [] to select ranges of characters to operate on

**Comment by kjg** [comment]
@sodabrew Is there any further discussion needed about use of `length` and such, or is this ready for merge since `length` is what has been used all along?

**Comment by kjg** [comment]
For what it's worth, I have my branch running in a production env, and it has processed 74,000 emails without issue. It has seen the same type of volume that used to trigger the stack level too deep errors, and it no longer does with this change.

**Comment by sodabrew** [comment]
I just read this over a couple times, and I think it's good. Thank you for the contribution!

**Comment by kjg** [comment]
Is there anything holding this up from getting merged, or do you only merge the PRs when preparing a release?

**Comment by kjg** [comment]
Is there anything holding this up from getting merged, or do you only merge the PRs when preparing a release?

**Comment by sodabrew** [comment]
Apologies for the long delay, I've had my head soaked in the mysql2 gem for its next big release. I'll find time for an eventmachine next minor merge/release as soon as possible.

**Comment by sodabrew** [comment]
I finally read this with `git show -w` and it's a clear and simple diff. Merging!

**Pull request merged**